keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/Stepsize.java

```diff
@@ -0,0 +1,165 @@
+        this.shrinkageTarget = Math.log(10 * stepsize);
+        this.averageTreeAcceptanceProb = 0.;
+    }
+
+    /**
+     * Taken from algorithm 4 in https://arxiv.org/pdf/1111.4246.pdf.
+     *
+     * @param position                      the starting position
+     * @param gradient                      the gradient at the starting position
+     * @param vertices                      the vertices
+     * @param probabilisticVertices         the probabilistic vertices
+     * @param logProbGradientCalculator     the log prob gradient calculator
+     * @param initialLogOfMasterP           the initial master log prob
+     * @param random                        the source of randomness
+     *
+     * @return a starting step size
+     */
+    public static double findStartingStepSize(Map<VertexId, DoubleTensor> position,
+                                       Map<VertexId, DoubleTensor> gradient,
+                                       List<Vertex<DoubleTensor>> vertices,
+                                       List<Vertex> probabilisticVertices,
+                                       LogProbGradientCalculator logProbGradientCalculator,
+                                       double initialLogOfMasterP,
+                                       KeanuRandom random) {
+        double stepsize = STARTING_STEPSIZE;
+        Map<VertexId, DoubleTensor> momentums = new HashMap<>();
+        initializeMomentumForEachVertex(vertices, momentums, random);
+
+        Leapfrog leapfrog = new Leapfrog(position, momentums, gradient);
+        double pThetaR = initialLogOfMasterP - leapfrog.halfDotProductMomentum();
+
+        Leapfrog delta = leapfrog.step(vertices, logProbGradientCalculator, STARTING_STEPSIZE);
+
+        double probAfterLeapfrog = ProbabilityCalculator.calculateLogProbFor(probabilisticVertices);
+        double pThetaRAfterLeapFrog = probAfterLeapfrog - delta.halfDotProductMomentum();
+
+        double logLikelihoodRatio = pThetaRAfterLeapFrog - pThetaR;
+        double scalingFactor = logLikelihoodRatio > Math.log(0.5) ? 1 : -1;
+
+        while (scalingFactor * logLikelihoodRatio > -scalingFactor * Math.log(2)) {
+            stepsize = stepsize * Math.pow(2, scalingFactor);
+
+            delta = leapfrog.step(vertices, logProbGradientCalculator, stepsize);
+            probAfterLeapfrog = ProbabilityCalculator.calculateLogProbFor(probabilisticVertices);
+            pThetaRAfterLeapFrog = probAfterLeapfrog - delta.halfDotProductMomentum();
+
+            logLikelihoodRatio = pThetaRAfterLeapFrog - pThetaR;
+        }
+
+        return stepsize;
+    }
+
+    /**
+     * Adapts the step size based on the state of the tree and network after computing a sample
+     *
+     * @param tree          the balanced binary tree
+     * @param sampleNum     the number of samples that have been taken
+     *
+     * @return a new step size
+     */
+    public double adaptStepSize(TreeBuilder tree, int sampleNum) {
```

Comment:
I would find it clearer if you pull out another method so that this becomes:

    public double adaptStepSize(TreeBuilder tree, int sampleNum) {

        double logStepSize = logStepSizeFrozen;

        if (sampleNum < adaptCount) {
            logStepSize = updateLogStepSize(tree, sampleNum);
        }

        stepsize = Math.exp(logStepSize);
        return stepsize;
    }